### ISSUE TITLE
RefineMesh lockup

### ISSUE THREAD

**Issue by YouYue123** [cdcseacave/openMVS#569]
### Specifications like the version of the project, operating system, and hardware
 Build date: May 22 2020, 10:24:57
 CPU: Intel(R) Xeon(R) CPU @ 2.00GHz (16 cores)
 RAM: 58.88GB Physical Memory 0B Virtual Memory
 OS: Linux 5.3.0-1020-gcp (x86_64)
 SSE & AVX compatible CPU & OS detected
### Steps to reproduce the problem

Running RefineMesh lock up into an infinity loop. It seems the gradient descent algorithm could not determine a converge in this situation. Any option I can use to resolve this? Thanks. Here is the log 

Decimated faces 207 (100%, 1ms)
08:53:42 [App     ] Cleaned mesh: 42 vertices, 80 faces (1ms)
08:53:42 [App     ] Cleaned mesh: 42 vertices, 80 faces (0ms)
08:53:42 [App     ] Cleaned mesh: 33 vertices, 62 faces (0ms)

**Comment by YouYue123** [comment]
It is related to the --max-threads setting. When I set the full 16 cores to do it, it will be stuck. 

Maybe some race condition happens. Could you give me any hint about which part of the code will possibly cause this? I am glad to contribute for further investigation and code commit

**Comment by cdcseacave** [comment]
I doubt that is the main issue; pls note your mesh has only 80 faces; you get a bad result in previous steps; use Viewer to check every step

**Comment by YouYue123** [comment]
Yes, the mesh is quite simple and not really complete. But at least RefineMesh should not be stuck and block the following script.

This issue only happens when I set **--max-threads > 8** in my environment. Will it be possible that there is a deadlock situation for simple mesh refinement?
